service/src/main/kotlin/io/provenance/digitalcurrency/consortium/service/CoinMintService.kt

```diff
@@ -16,29 +18,40 @@ class CoinMintService(
 ) {
     private val log = logger()
 
-    fun createEvent(coinMintRecord: CoinMintRecord) {
-        // There should not be any tx events at this point or all event should have an error status
-        val existingEvents: List<TxStatusRecord> =
-            TxStatusRecord.findByTxRequestUuid(coinMintRecord.id.value)
-        check(
-            existingEvents.isEmpty() ||
-                existingEvents.filter { it.status == TxStatus.ERROR }.size == existingEvents.size
-        ) { "Mint/swap contract already called" }
-
-        try {
-            val txResponse = pbcService.mintAndSwap(
-                amount = coinMintRecord.coinAmount.toBigInteger(),
-                address = coinMintRecord.addressRegistration.address
-            ).txResponse
-            log.info("Mint/swap tx hash: ${txResponse.txhash}")
-            TxStatusRecord.insert(
-                txResponse = txResponse,
-                txRequestUuid = coinMintRecord.id.value,
-                type = TxType.MINT_CONTRACT
-            )
-            CoinMintRecord.updateStatus(coinMintRecord.id.value, CoinMintStatus.PENDING_MINT)
-        } catch (e: Exception) {
-            log.error("Mint/swap contract failed; it will retry.", e)
+    fun createEvent(coinMintRecords: List<CoinMintRecord>) {
+        coinMintRecords.mapNotNull { coinMintRecord ->
+            // There should not be any tx events at this point or all event should have an error status
+            val existingEvents: List<TxStatusRecord> =
+                TxStatusRecord.findByTxRequestUuid(coinMintRecord.id.value)
+            if (!(existingEvents.isEmpty() || existingEvents.filter { it.status == TxStatus.ERROR }.size == existingEvents.size)) {
+                log.error("Mint contract already called")
+                CoinMintRecord.updateStatus(coinMintRecord.id.value, CoinMintStatus.EXCEPTION)
+                null
+            } else {
+                ExecuteMintRequest(
+                    mint = MintRequest(
+                        amount = coinMintRecord.coinAmount.toString(),
+                        address = coinMintRecord.addressRegistration.address
+                    )
+                )
+            }
+        }.takeIf { mintMessages: List<ExecuteMintRequest> ->
+            mintMessages.isNotEmpty()
+        }?.let { mintMessages ->
+            try {
+                val txResponse = pbcService.mintBatch(mintMessages).txResponse
```

Comment:
although we're batching mints, don't we still have a problem with then multiple actors competing for tx broadcasts in different threads? its no longer one tx per block but its one tx per type which can still run into sequence number issues?

Reply:
might be better to have a single actor/corouting query all the blockchain requests asc and handle the various types of tx's together?

also due to timing it might be safer to while loop/thread sleep if there are any pending tx hashes not processed yet before getting another batch to fully avoid sequence number errors